sf-stream/src/runtime.rs

```diff
@@ -0,0 +1,232 @@
+            mp_sender: mp_client_sender,
+        }
+    }
+
+    pub async fn start(&mut self) {
+        loop {
+            self.batch_convert(100).await;
+        }
+    }
+
+    pub async fn batch_convert(&mut self, batch_size: u16) -> Vec<extractor::Transaction> {
+        let mut result: Vec<extractor::Transaction> = vec![];
+        match &self.context.db.get_first_txn_version() {
+            Ok(version_result) => match version_result {
+                Some(oldest_version) => {
+                    if oldest_version > &self.current_version {
+                        warn!(
+                            "[sf-stream] oldest txn version is {} but requested version is {}",
+                            oldest_version, &self.current_version
+                        );
+                        sleep(Duration::from_millis(300)).await;
+                        return vec![];
+                    }
+                }
+                None => {
+                    return vec![];
+                },
+            },
+            Err(err) => {
+                warn!("[sf-stream] failed to get first txn version: {}", err);
+                sleep(Duration::from_millis(300)).await;
+                return vec![];
+            }
+        };
+
+        let ledger_info = self.context.get_latest_ledger_info().unwrap();
+        match self
+            .context
+            .get_transactions(self.current_version, batch_size, ledger_info.version())
+        {
+            Ok(transactions) => {
+                if transactions.is_empty() {
+                    debug!("[sf-stream] no transactions to send");
+                    sleep(Duration::from_millis(100)).await;
+                    return result;
+                }
+                // TODO: there might be an off by one (tx version fetched)
+                debug!(
+                    "[sf-stream] got {} transactions from {} to {} [{}]",
+                    transactions.len(),
+                    self.current_version,
+                    self.current_version + transactions.len() as u64,
+                    transactions.last().map(|txn| txn.version).unwrap_or(0)
+                );
+                for onchain_txn in transactions {
+                    // TODO: assert txn.version == &self.current_version + 1?
+                    // TODO: return a `Result` & check to ensure we pushed before incrementing
+                    let txn_version = onchain_txn.version;
+                    // Todo: since the timestamp is per block, only calculate this value once per block
+                    let timestamp = self
+                        .context
+                        .get_block_timestamp(onchain_txn.version)
+                        .unwrap_or_else(|_| {
+                            panic!(
+                                "Could not get timestamp for version {}",
+                                onchain_txn.version
+                            )
+                        });
```

Comment:
can you just use `expect` here?

Reply:
I wanted to say this a couple places, but it's an annoying thing with expect, that you can't do a format string, so you have to use an `unwrap_or_else`